# HG changeset patch
# User Markus Muetzel <markus.muetzel@gmx.de>
# Date 1468253305 -7200
#      Mon Jul 11 18:08:25 2016 +0200
# Branch stable
# Node ID fba8a6c742c6346261ca4e1735f903f59ce0772f
# Parent  8defe99c96508628d44a2ae4e74a709b0a1608cb
fminsearch.m: Fix error in "guarded_eval".

* fminsearch.m: Fix typo in guarded_eval which meant nothing was being tested.
Add BIST tests to check for proper behavior.

diff --git a/scripts/optimization/fminsearch.m b/scripts/optimization/fminsearch.m
--- a/scripts/optimization/fminsearch.m
+++ b/scripts/optimization/fminsearch.m
@@ -324,30 +324,40 @@ function [x, fmax, nf] = nmsmax (fun, x,
 
 endfunction
 
 ## A helper function that evaluates a function and checks for bad results.
 function y = guarded_eval (fun, x)
 
   y = fun (x);
 
-  if (! (isreal (f)))
+  if (! (isreal (y)))
     error ("fminsearch:notreal", "fminsearch: non-real value encountered");
-  elseif (any (isnan (f(:))))
+  elseif (any (isnan (y(:))))
     error ("fminsearch:isnan", "fminsearch: NaN value encountered");
-  elseif (any (isinf (f(:))))
+  elseif (any (isinf (y(:))))
     error ("fminsearch:isinf", "fminsearch: Inf value encountered");
   endif
 
 endfunction
 
 
 %!demo
-%! fcn = @(x) (x(1)-5).^2 + (x(2)-8).^4
+%! fcn = @(x) (x(1)-5).^2 + (x(2)-8).^4;
 %! x0 = [0;0];
 %! [xmin, fval] = fminsearch (fcn, x0)
 
-%!assert (fminsearch (@sin, 3, optimset ("MaxIter", 3)), 4.8750, 1e-4)
-%!assert (fminsearch (@sin, 3, optimset ("MaxIter", 30)), 4.7124, 1e-4)
-%!shared c
+%!assert (fminsearch (@sin, 3, optimset ("MaxIter", 30)), 3*pi/2, 1e-4)
+%!test
 %! c = 1.5;
-%!assert (fminsearch (@(x) x(1).^2+c*x(2).^2,[1;1]), [0;0], 1e-4)
+%! assert (fminsearch (@(x) x(1).^2+c*x(2).^2,[1;1]), [0;0], 1e-4);
+
+%!error fminsearch ()
+%!error fminsearch (1)
 
+## Tests for guarded_eval
+%!error <non-real value encountered>
+%! fminsearch (@(x) ([0 2i]), 0, optimset ("FunValCheck", "on"));
+%!error <NaN value encountered>
+%! fminsearch (@(x) (NaN), 0, optimset ("FunValCheck", "on"));
+%!error <Inf value encountered>
+%! fminsearch (@(x) (Inf), 0, optimset ("FunValCheck", "on"));
+
